src/finiteVolume/processes/surfaceForce.cpp

```diff
@@ -0,0 +1,303 @@
+        // get current euler solution here to get velocity
+        const PetscScalar *euler = nullptr;
+        DMPlexPointLocalFieldRead(dm, c, eulerField.id, solArray, &euler);
+        auto density = euler[ablate::finiteVolume::CompressibleFlowFields::RHO];
+
+        PetscScalar vel[dim];
+        for (PetscInt d = 0; d < dim; d++) {
+            vel[d] = euler[ablate::finiteVolume::CompressibleFlowFields::RHOU + d] / density;
+        }
+
+        // add calculated sources to euler
+        PetscScalar *eulerSource = nullptr;
+        DMPlexPointLocalFieldRef(dm, c, eulerField.id, fArray, &eulerSource);
+
+        PetscReal surfaceEnergy = 0;
+        PetscReal totalGradMagNormal = 0;
+        PetscReal divergentNormal[3] = {0, 0, 0};
+        PetscReal grad[3] = {0, 0, 0};
+        PetscReal gradMagNormal[dim];
+        PetscReal curvature;
+        PetscScalar surfaceForce[dim];
+
+        // get the centroid information for the cell
+        PetscFVCellGeom *fcg;
+        DMPlexPointLocalRead(dmCell, c, cellGeomArray, &fcg);
+
+        // extract connected vortices to the cell
+        PetscInt *closure = NULL;
+        PetscInt numClosure;
+        DMPlexGetTransitiveClosure(dm, c, PETSC_TRUE, &numClosure, &closure);
+        PetscInt cl, numVertex = 0;
+
+        PetscReal centerNormal[3] = {0, 0, 0};
+        PetscReal gradNormal[dim];
+        PetscReal cellCenterNormal[dim];
+        PetscReal totalDivNormal = 0;
+        PetscInt vStart, vEnd;
+        DMPlexGetDepthStratum(dm, 0, &vStart, &vEnd);
+        PetscReal distance[3] = {0, 0, 0};
+
+        // extract the local x array
+        PetscReal xyz[3];
+        PetscInt offset;
+
+        for (cl = 0; cl < numClosure * 2; cl += 2) {
+            PetscInt vertex = closure[cl];
+
+            if (vertex < vStart || vertex >= vEnd) continue;
+            // sum up the number of connected vortices
+            numVertex += 1;
+            PetscSectionGetOffset(coordsSection, vertex, &offset);
+
+            for (PetscInt d = 0; d < dim; ++d) {
+                // extract coordinates of vortices of this cell and calculate the distance to the center
+                DMPlexPointLocalRead(cdm, vertex, normalArray, &vertexNormal);
+                xyz[d] = coordsArray[offset + d];
+
+                // calculate normal at each cell center using normals of connected vortices to the cell
+                centerNormal[d] += vertexNormal[d];
+
+                // calculate divergence of normal for each cell center using vertex normals --> Delta.ni,j=  [nx(i+1/2,j+1/2)-nx(i-1/2,j+1/2)+nx(i+1/2,j-1/2)-nx(i-1/2,j-1/2)]/2*deltaX +
+                // [ny(i+1/2,j+1/2)-ny(i+1/2,j-1/2)+ny(i-1/2,j+1/2)-ny(i-1/2,j-1/2)]/2*deltaY
+                // get magnitude of normals at vortices to calculate the derivative of the magnitude of normals at the cell center
+                distance[d] += abs(xyz[d] - fcg->centroid[d] + utilities::Constants::tiny);
```

Comment:
Couldn't this distance calculation be done in the stencil so you would not need to get xyz each time?

Reply:
I am only marching over cells here, not in the setup, so probably just do it here.
Regarding close cells line up, I checked the code again, and I calculated weights everywhere that gradients are tasked as the original ALE- scheme suggests. So I believe there may not be any fake gradients or singularity problems.
....
Still waiting to hear back from Dr. Keneply!